src/IROperator.cpp

```diff
@@ -248,13 +273,62 @@ Expr const_false(int w) {
     return make_zero(UInt(1, w));
 }
 
+Expr lossless_cast(Type t, Expr e) {
+    if (t == e.type()) {
+        return e;
+    } else if (t.can_represent(e.type())) {
+        return cast(t, e);
+    }
+
+    if (const Cast *c = e.as<Cast>()) {
+        if (t == c->value.type()) {
+            return c->value;
+        } else {
+            return lossless_cast(t, c->value);
+        }
+    }
+
+    if (const Broadcast *b = e.as<Broadcast>()) {
+        Expr v = lossless_cast(t.element_of(), b->value);
+        if (v.defined()) {
+            return Broadcast::make(v, b->width);
+        } else {
+            return Expr();
+        }
+    }
+
+    if (const IntImm *i = e.as<IntImm>()) {
+        if (t.can_represent(i->value)) {
+            return IntImm::make(t, i->value);
```

Comment:
It t is UInt this will assert inside IntImm::make won it? (Ditto for UInt case below.) I'm pretty sure this can happen with the right set of parameters to this routine.